### DISCUSSION TITLE
是不是应该移除组件库样式？

### DISCUSSION THREAD

**Discussion by janryWang** [alibaba/formily#1501] [Ideas]
现在在每个组件style.ts里都依赖了对应组件库的样式，感觉不太合理，很容易重复打包，如果改为仅仅只是在style.less文件里依赖组件库主题变量，mixins，这样会不会更好，对应组件库的样式文件由业务侧自行引入

**Comment by liuweiGL** [comment]
我看 [ant-design/pro-components](https://github.com/ant-design/pro-components) 就是只引入 `themes/default.less` 变量，antd 组件库的样式文件自己处理。

**Reply by imsobear** [reply]
最终 js 里是引了的：

![image](https://user-images.githubusercontent.com/2505411/119468429-05be9280-bd79-11eb-9ebd-7999061414eb.png)

**Comment by liuweiGL** [comment]
还有就是入口文件不需要自动引入 style 文件，这个操作也应该交给用户。

**Reply by liuweiGL** [reply]
因为如果我只使用了 @formily/antd 包中的某几个组件，但是你却把所有组件的样式都给我导入进来，这应该是不合理的操作。

**Comment by djknight1** [comment]
应该去掉，目前的方案会有以下几个问题：
1. formily 包体积变大；
2. 样式会被覆盖；如果有其它三方 framework 依赖相同的组件库，并定义好了主题样式，此主题样式会被 formily 覆盖；
3. 从心智上讲，formily 作为一个 lib，不应该帮用户导入所有的样式，不合理；

**Comment by janryWang** [comment]
移除组件库样式依赖应该是合理的，就是应该会有break change，打算beta.58上尝试一下，看看会不会有其他问题

**Reply by djknight1** [reply]
有发布计划吗？ 58上好像还没看到

**Comment by imsobear** [comment]
移除之后，如果用户的代码是 `import { Input } from '@formily/antd-components';`，对应的 `antd/es/input/style` 是需要开发者自己引入？

**Reply by janryWang** [reply]
讲道理，用户配置了ts-import-plugin的话，应该是可以直接把formily内部依赖的组件，自动引入对应组件样式的，因为formily内部是走的import { Button } from "antd"这种写法，如果是全量引入formily包的话，应该得在项目维度单独引入antd.css或者antd.less

**Reply by liuweiGL** [reply]
你使用 antd 的时候肯定都配好了按需加载或则引入整儿 antd.css。

**Reply by imsobear** [reply]
全量引入的场景不需要讨论，只讨论按需场景。

按需大部分是配置 babel-plugin-import，然后 babel-loader 大部分时候是 exclude node_modules 的，因此这个情况下 babel 是分析不到依赖里面（比如 formily）的代码的。

**Reply by liuweiGL** [reply]
ant-design/pro-components 这样相当于埋雷，我如果引入了 antd.css，它给我整个这个搞的样式覆盖，用户排查起来一万个草泥马。源码不写明，在 build 阶段来搞也是 6。

**Reply by imsobear** [reply]
在我们这边，全量引入本身就不是推荐的方案。另外这里不讨论全量引入的场景。

**Reply by liuweiGL** [reply]
> 全量引入的场景不需要讨论，只讨论按需场景。
> 
> 按需大部分是配置 babel-plugin-import，然后 babel-loader 大部分时候是 exclude node_modules 的，因此这个情况下是失效的。

那可以在文档加上提醒，用户手动创建一个 less 文件来集中引入或则修改 loader 配置都行。

**Reply by liuweiGL** [reply]
> 在我们这边，全量引入本身就不是推荐的方案。另外这里不讨论全量引入的场景。

那不代表所有的人都不使用全量引入，如果你在类库里面自动引入，使用 antd.css 的用户就没办法处理。

**Reply by janryWang** [reply]
你们如果是exclude node_modules，按需打包场景是如何做到不自动加载formily组件的style.js的呢，因为组件js里没有import这个文件

**Reply by imsobear** [reply]
> > 全量引入的场景不需要讨论，只讨论按需场景。
> > 按需大部分是配置 babel-plugin-import，然后 babel-loader 大部分时候是 exclude node_modules 的，因此这个情况下是失效的。
> 
> 那可以在文档加上提醒，用户手动创建一个 less 文件来集中引入或则修改 loader 配置都行。

所以结论就是需要手动引入了，比如：

``` js
import { Input } from '@formily/next-components';
import 'antd/es/Input/style';
import 'formily/next-components/es/Input/style';  // 这个可以通过 babel-plugin-import 自动引入，如果没有覆盖样式的话也不需要引入
```

**Reply by imsobear** [reply]
> 你们如果是exclude node_modules，按需打包场景是如何做到不自动加载formily组件的style.js的呢，因为组件js里没有import这个文件

- exclude node_modules 已经是一个最佳实践了，大部分框架现在都是这么做的。
- 关于 formily/antd-components 的使用，我的推荐是把 ta 作为一个大的组件包，跟 antd 保持一致配置一个 babel-plugin-import，自动引入对应组件的 style.js，然后对应的 antd 组件 style.js 也就会引进来了，但是如果 style.js 里移除了基础组件的 style.js 的话这个方案就有问题了

**Reply by janryWang** [reply]
> > 你们如果是exclude node_modules，按需打包场景是如何做到不自动加载formily组件的style.js的呢，因为组件js里没有import这个文件
> 
> * exclude node_modules 已经是一个最佳实践了，大部分框架现在都是这么做的。
> * 关于 formily/antd-components 的使用，我的推荐是把 ta 作为一个大的组件包，跟 antd 保持一致配置一个 babel-plugin-import，自动引入对应组件的 style.js，然后对应的 antd 组件 style.js 也就会引进来了，但是如果 style.js 里移除了基础组件的 style.js 的话这个方案就有问题了

就是感觉这种用法有点定制了，因为外部很多用户基本上都是用的umi或者纯webpack，umi没有对formily做特殊处理，这里就会存在一个gap，使用babel-import-plugin(如果exclude node_modules了)是无法自动引入antd样式的，需要自己再写个工程工具去导入formily 组件包里的style.js

**Reply by imsobear** [reply]
icejs 也没有对 formily 特殊处理，没特殊处理是正常的，问题是这种情况你这边推荐的最佳实践是什么？

**Reply by janryWang** [reply]
那我的问题就是，为啥没有特殊处理，babel-import-plugin就能自动加载formily组件的style.js，因为formily组件的主代码是没有import这个style.js的

**Reply by imsobear** [reply]
> 那我的问题就是，为啥没有特殊处理，babel-import-plugin就能自动加载formily组件的style.js，因为formily组件的主代码是没有import这个style.js的

babel-plugin-import 自动加载了 formily/antd-components 对应组件比如 Input 的 style.js 这个没问题呀，问题是你这个 Input 背后依赖的 antd Input 的 style.js 谁去加载，怎么加载？

**Reply by liuweiGL** [reply]
最佳实践也许是全量引入：https://github.com/ant-design/babel-plugin-import/issues/369#issuecomment-551636048 😃

**Comment by liuweiGL** [comment]
主题描述可能有点不明确，我们讨论的应该是 formily next 版本。

**Comment by liuweiGL** [comment]
在这里也补充一下 vite 里面按需加载存在的问题吧。

## 问题描述

第三方框架中引入 antd 组件，使用 vite 的时候按需加载无效。

## 原因

vite 有个 pre-compile 环节，会把 node_modules 中的包预先 bundle 一下，减少 dev server 中的 http 请求数量。但是 bundle 的时候会把  `import` 语句破坏掉：
```ts
import {
  badge_default,
  button_default,
  card_default,
  cascader_default,
  checkbox_default,
  collapse_default,
  date_picker_default,
  drawer_default,
  input_default,
  input_number_default,
  modal_default,
  pagination_default,
  popover_default,
  radio_default,
  select_default,
  space_default,
  steps_default,
  switch_default,
  table_default,
  tabs_default,
  tag_default,
  time_picker_default,
  transfer_default,
  tree_select_default,
  upload_default
} from "./chunk-SJEL7XUF.js";
```

可以看到，本来是 `from 'antd'` 的，结果变成了 `./chunk-SJEL7XUF.js`，因为 antd 被 bundle 成这个文件了。

**Comment by liuweiGL** [comment]
不管怎么说，让用户来决定样式文件的加载时机是合理的，其中可能有些坑，我们发现之后补充到文档就够了。

**Comment by janryWang** [comment]
确定下来方案了，下面明确一下：
1. 支持按需打包的模式保持不变，同样是每个组件目录下都有一个style.ts，引用写法不变
2. 非按需打包模式的@formily/antd/{lib,esm}/style.ts中不依赖每个组件的style.ts文件，改为只依赖组件的less文件，这样可以保证普通用法下绝对不会将antd整个包的样式打进去，按需打包模式下不会受任何影响

**Reply by liuweiGL** [reply]
@formily/antd 根目录下的 index.ts 还会引入 `./style.ts` 吗？

**Reply by janryWang** [reply]
会的，只是style.ts里的内容只有自身样式和antd的主题变量

**Reply by liuweiGL** [reply]
为什么不选择在每个组件各自的 index.ts 中 import 自身的样式，而选择在包的入口 index.ts 引入整个 style 呢？

<details>
<summary>感觉应该这样</summary>
 <img src="https://user-images.githubusercontent.com/21048878/119509484-4206e880-bda3-11eb-84d7-215ad3808993.png"/>

</details>

<details>
<summary>当前是这样的</summary>
<img src="https://user-images.githubusercontent.com/21048878/119509606-5e0a8a00-bda3-11eb-80b2-f1c4b8a446ba.png" />
</details>

**Reply by liuweiGL** [reply]
在包入口集中引入就会导致上面我说的，我只用到了 Input，结果把所有组件的 style 都打包进来了。

**Reply by janryWang** [reply]
> 为什么不选择在每个组件各自的 index.ts 中 import 自身的样式，而选择在包的入口 index.ts 引入整个 style 呢？
> 
> 感觉应该这样
> 当前是这样的

会和按需打包的冲突

**Reply by janryWang** [reply]
> 在包入口集中引入就会导致上面我说的，我只用到了 Input，结果把所有组件的 style 都打包进来了。

你都是按需打包了，就会走按需打包逻辑，互不影响的

**Reply by liuweiGL** [reply]
奥，是的，我反应过来了，按需加载不会执行到包的入口文件。

**Reply by liuweiGL** [reply]
[@formily_antd.zip](https://github.com/alibaba/formily/files/6550796/%40formily_antd.zip)

入口的 index 里面不能 import style，不然 vite pre-compile 的时候会把所有 less 文件都 import 进来。

**Reply by janryWang** [reply]
你说的是按需打包还是非按需打包

**Reply by liuweiGL** [reply]
> > 为什么不选择在每个组件各自的 index.ts 中 import 自身的样式，而选择在包的入口 index.ts 引入整个 style 呢？
> > 感觉应该这样
> > 当前是这样的
> 
> 会和按需打包的冲突

我想了下，不会冲突，只需要 babel-import-plugin 不要配置 @formily/core 就好了啊：

```ts
import { ArrayTable } from '@formily/antd'
```
按需加载的时候变成：
```ts
import ArrayTable from '@formily/antd/es/ArrayTable'
```

这个时候 ArrayTable.ts 里面自动引入了样式， babel-import-plugin 就不要配置了啊。退一步讲，就算配置了也没事吧，因为 less 是会去重的。

**Reply by janryWang** [reply]
现在已经发包了，按照确定的方案来了，你先试试有啥问题再讨论吧

**Comment by liuweiGL** [comment]
使用 vite 的时候，如果每个第三方类库都导入 antd 组件的样式，那么同一组件的样式是会被引入多次的。

demo：https://github.com/liuweiGL/formily-style-conflicted

效果：
![image](https://user-images.githubusercontent.com/21048878/119505131-3d403580-bd9f-11eb-8805-738919c44fe0.png)

**Comment by yee94** [comment]
我认为是有必要移除组件所用的样式的，Formily如果本身自己有一些样式定义自然是可以引入，但是没必要再引入组件库本身的样式。

Fusion中，在引入组件的过程中本身就有对组件样式的按需引入配置方法。

Formily 在组件使用过程中使用`import "@alifd/next/lib/input/style.js"`的方式，在直接使用时是不透明的，这很可能会导致在使用到`@formily/next`的用户重复打包组件样式，对于一些External引入Fusion主题的用户来说，甚至会出现主题冲突的问题。

个人认为，组件本身的样式`import "@alifd/next/lib/input/style.js";`应该交由组件本身import去处理`import { Input } from "@alifd/next";`， 如果Formily定制化的组件使用到组件库的变量可以仅引入 variables.scss ，但最好不要单独特定组件样式。

**Comment by liuweiGL** [comment]
另开一层楼吧，我建个 demo 先整明白 vite 怎么编译的。
